docs/decisions/0021-transaction-aggregates.rst

```diff
@@ -0,0 +1,135 @@
+
+Context
+=======
+Part of the primary responsibility of the ``subsidy_access_policy`` module is to
+act as a "read-orchtestrator" for redeemability, determining which learner-credit-based redemptions
+are allowed under various conditions. The business logic to determine this depends on
+data persisted elsewhere in our distributed system - importantly, it relies on the
+enterprise-subsidy to compute aggregate transaction quantity sums (or "spend") at various levels:
+for the entire subsidy/ledger record, for the policy (or "budget"), and for the learner within
+a given policy. There are also forthcoming circumstances where the enterprise-access service
+would like to know the aggregate *count* of learner transactions for a given policy-group identifier.
+
+This distributed dependency design was chosen because it helps decouple services/responsibilities
+from a a deployment standpoint, and it helps isolate/secure the transaction-level data in the
+enterprise-subsidy service. However, this design has several tradeoffs:
+
+1. Tight "runtime" dependencies exist - the ``can_redeem`` business logic and the redemption
+   workflow cannot execute at all without both the enterprise-access and enterprise-subsidy services
+   running. Furthermore, that data in each service is highly coupled - policy records in enterprise-access
+   *must*  relate to a well-formed and active *subsidy* record in the enterprise-subsidy service in
+   order for the redemption workflow to execute.
+2. These runtime dependencies utilize synchronous HTTP requests for both the ``can_redeem`` and
+   ``redeem`` workflows. This makes the system as a whole less scalable, less elastic,
+   and less performant, because each of these workflows mush synchronously block on requests from
+   enterprise-access to enterprise-subsidy.
+   
+
+Decision
+========
+Build and maintain *aggregate* redemption models in the enterprise-access service. The aggregated
+metrics we're most interested in right now are total spend (the sum of transaction quantities) and the count
+of transactions.
+The aggregated *dimensions* we're most interested in right now are as follows:
+
+1. count/spend metrics at the **subsidy** level - based on ``subsidy_uuid``
+2. count/spend metrics at the **policy** level - based on ``subsidy_access_policy_uuid``
+3. count/spend metrics at the **(policy, learner)** level - based on (policy id, learner id)
+
+Additionally, we may eventually want aggregates at the **(policy, learner, group)** level, to support
+business record configurations where multiple groups relate to a single policy.
+
+Once these aggregates are in place, replace requests to the enterpirse-subsidy service
+in the ``can_redeem`` and ``redeem`` flows with database queries against these aggregates.
+
+Maintaining these aggregates depends on sourcing transaction and reversal data as follows:
+
+1. Starting from a zero-state in which Aggregate models exist but are not yet populated (backpopulation).
```

Comment:
I think we should also call out recalculation here too.

Reply:
Good call, added a note about this, specifically for use in error correction.